### PR TITLE
🎉 Support incremental sync in Sendgrid

### LINKED ISSUE #1382: Support incremental sync in Sendgrid
## Tell us about the problem you're trying to solve
Blocked by #3255 
The end goal of this ticket is make the Sendgrid source support incremental sync on all streams which support it. 

Before beginning implementation we should: comment on this ticket noting which streams support incremental sync. 

## Acceptance Criteria
1. Sendgrid source supports incremental sync
1. docs are updated to indicate this
1. Use new standard test suite for testing

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368079125479) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by htrueman**
## What
Added incremental streams support to `global_suppressions`, `blocks`, `bounces`, `invalid_emails` and `spam_reports` streams.

## How
* Added incremental configured catalogs json configs
* Updated source to support `start_time` config
* Added `SendgridStreamIncrementalMixin` to be inherited by streams that support incremental sync

## Pre-merge Checklist
* [ ] _Run integration tests_
* [ ] _Publish Docker images_

## Recommended reading order
1. Configured catalogs json files
1. `source.py`
1. `streams.py`

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368195282526) by [Unito](https://www.unito.io)

### CODE REVIEW

File: airbyte-integrations/connectors/source-sendgrid/sample_files/configured_catalog.json [hunk 1]
```diff
@@ -291,10 +291,12 @@
             }
           }
         },
```
Comment (keu) [inline airbyte-integrations/connectors/source-sendgrid/sample_files/configured_catalog.json:293]: we don't need to have two configured catalogs, lets leave only one in integration_tests

File: airbyte-integrations/connectors/source-sendgrid/source_sendgrid/streams.py [hunk 2]
```diff
@@ -75,6 +76,33 @@ def next_page_token(self, response: requests.Response) -> Optional[Mapping[str,
         return {"offset": self.offset}
 
 
+class SendgridStreamIncremental(SendgridStream):
```
Comment (keu) [inline airbyte-integrations/connectors/source-sendgrid/source_sendgrid/streams.py:79]: no usage of this

File: airbyte-integrations/connectors/source-sendgrid/source_sendgrid/streams.py [hunk 3]
```diff
@@ -75,6 +76,33 @@ def next_page_token(self, response: requests.Response) -> Optional[Mapping[str,
...
+    pass
+
+
+class SendgridStreamIncrementalMixin(HttpStream):
+    cursor_field = "created"
+
+    def __init__(self, start_time, **kwargs):
```
Comment (keu) [inline airbyte-integrations/connectors/source-sendgrid/source_sendgrid/streams.py:86]: can we have type annotations please

File: airbyte-integrations/connectors/source-sendgrid/source_sendgrid/streams.py [hunk 4]
```diff
@@ -75,6 +76,33 @@ def next_page_token(self, response: requests.Response) -> Optional[Mapping[str,
...
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+        latest_benchmark = latest_record[self.cursor_field]
+        if current_stream_state.get(self.cursor_field):
+            return {self.cursor_field: max(latest_benchmark, current_stream_state[self.cursor_field])}
+        return {self.cursor_field: latest_benchmark}
```
Comment (keu) [inline airbyte-integrations/connectors/source-sendgrid/source_sendgrid/streams.py:98]:
```suggestion
        value = max(latest_benchmark, current_stream_state.get(self.cursor_field, latest_benchmark)
        return {self.cursor_field: value}
```

**Review by keu — APPROVED** [review]
LGTM, just a small comment, otherwise approved, did you enable acceptance tests for incremental?

File: airbyte-integrations/connectors/source-sendgrid/sample_files/configured_catalog.json (hunk 1 above)
Comment (htrueman) [reply airbyte-integrations/connectors/source-sendgrid/sample_files/configured_catalog.json]: Integration tests fail if not sample_files/configured_catalog.json exists
```
    def _opener(self, name, flags, mode=0o666):
        # A stub for the opener argument to built-in open()
>       return self._accessor.open(self, flags, mode)
E       FileNotFoundError: [Errno 2] No such file or directory: '/test_input/sample_files/configured_catalog.json'
```
May I put it back, or perhaps it better to remove it from `integrations_tests` instead?

**Comment by htrueman** [comment]
> LGTM, just a small comment, otherwise approved, did you enable acceptance tests for incremental?

It's added for now

File: airbyte-integrations/connectors/source-sendgrid/source_sendgrid/streams.py [hunk 5]
```diff
@@ -75,6 +76,29 @@ def next_page_token(self, response: requests.Response) -> Optional[Mapping[str,
         return {"offset": self.offset}
 
 
+class SendgridStreamIncrementalMixin(HttpStream):
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-sendgrid/source_sendgrid/streams.py:79]: nifty usage of mixin

**Review by sherifnada — APPROVED** [review]

**Pull request merged**